source/extensions/filters/network/rocketmq_proxy/active_message.cc

```diff
@@ -0,0 +1,289 @@
+    connection_manager_.stats().pop_message_active_.dec();
+  } break;
+  case RequestCode::SEND_MESSAGE: {
+    connection_manager_.stats().send_message_v1_active_.dec();
+  } break;
+  case RequestCode::SEND_MESSAGE_V2: {
+    connection_manager_.stats().send_message_v2_active_.dec();
+  } break;
+  case RequestCode::GET_ROUTE_INTO_BY_TOPIC: {
+    connection_manager_.stats().get_topic_route_active_.dec();
+  } break;
+  default:
+    break;
+  }
+}
+
+void ActiveMessage::createFilterChain() { router_ = connection_manager_.config().createRouter(); }
+
+void ActiveMessage::createFilterChain(absl::string_view target_host) {
+  router_ = connection_manager_.config().createRouter(target_host);
+}
+
+void ActiveMessage::sendRequestToUpstream() {
+  if (!router_) {
+    createFilterChain();
+  }
+
+  router_->sendRequestToUpstream(*this);
+}
+
+void ActiveMessage::sendRequestToUpstream(absl::string_view target_host) {
+  if (!router_) {
+    createFilterChain(target_host);
+  }
+  router_->sendRequestToUpstream(*this);
+}
+
+std::string ActiveMessage::clusterName() {
+  CommandCustomHeader* customHeader = request_->customHeader().get();
+  auto routeCommandCustomHeader = dynamic_cast<RoutingCommandCustomHeader*>(customHeader);
+  if (routeCommandCustomHeader) {
+    return routeCommandCustomHeader->topic();
+  } else {
+    ENVOY_LOG(error, "Cannot figure out cluster name from request header. Opaque: {}, Code: {}",
+              request_->opaque(), request_->code());
+  }
+  return std::string();
+}
+
+bool ActiveMessage::isOneway() { return request_->is_oneway(); }
+
+void ActiveMessage::onError(absl::string_view error_message) {
+  connection_manager_.onError(request_, error_message);
+}
+
+RemotingCommandPtr& ActiveMessage::downstreamRequest() { return request_; }
+
+void ActiveMessage::sendResponseToDownstream() {
+  if (response_) {
+    response_->opaque(request_->opaque());
+    connection_manager_.sendResponseToDownstream(response_);
+  }
+}
+
+void ActiveMessage::fillBrokerData(std::vector<BrokerData>& list, const std::string& cluster,
+                                   const std::string& brokerName, int64_t brokerId,
+                                   const std::string& address) {
+  bool found = false;
+  for (auto& entry : list) {
+    if (entry.cluster() == cluster && entry.brokerName() == brokerName) {
+      found = true;
+      if (entry.brokerAddresses().find(brokerId) != entry.brokerAddresses().end()) {
+        ENVOY_LOG(warn, "Duplicate broker_id found. Broker ID: {}, address: {}", brokerId, address);
+        continue;
+      } else {
+        entry.brokerAddresses()[brokerId] = address;
+      }
+    }
+  }
+
+  if (!found) {
+    std::unordered_map<int64_t, std::string> addresses;
+    addresses[brokerId] = address;
+
+    list.emplace_back(BrokerData(cluster, brokerName, std::move(addresses)));
+  }
+}
+
+void ActiveMessage::onQueryTopicRoute() {
+  std::string cluster_name = clusterName();
```

Comment:
Can you give an example of the metadata structure

Reply:
```proto
message BrokerMeta {
    int32 read_queue_num = 1;
    int32 write_queue_num = 2;
    string cluster_name = 3;
    string broker_name = 4;
    int32 broker_id = 5;
    int32 perm = 6;
}

message QuotaMeta {
    // key is version (e.g. "v1"), value is quota (e.g. "0.5", "0.5,0.7")
    map<string, string> quota_info = 1;
}

message BrokersInfo {
    // key is broker_addr (ip + ":" + port)
    map<string, BrokerMeta> brokers_meta = 1;
    // key is consumerGroup + "@" + topic
    map<string, QuotaMeta> quotas_meta = 2;
}
```
metadata's structure is <"envoy.filters.network.rocketmq_proxy", BrokersInfo>

Reply:
`QuotaMeta` has been removed from `filter_metadata`.